src/Container/GenericContainer.php

```diff
@@ -171,51 +153,62 @@ private function autowire(string $className, ContainerLog $log): object
+            // trying to build it.
+            ? $reflectionClass->newInstanceWithoutConstructor()
+
+            // Otherwise, use our autowireDependencies helper to automagically
+            // build up each parameter.
+            : $reflectionClass->newInstanceArgs(
+                $this->autowireDependencies($constructor)
+            );
+    }
+
+    /**
+     * @return ReflectionParameter[]
+     */
+    private function autowireDependencies(ReflectionMethod $method, array $parameters = []): array
+    {
         // Build the class by iterating through dependencies
         // and resolving them.
         $dependencies = [];
 
-        foreach ($reflectionClass->getConstructor()->getParameters() as $parameter) {
-            $dependencies[] = $this->autowireDependency($parameter, $log);
+        foreach ($method->getParameters() as $parameter) {
+            $dependencies[] = $this->autowireDependency(
+                $parameter,
+                $parameters[$parameter->getName()] ?? null
+            );
         }
 
-        return $reflectionClass->newInstanceArgs($dependencies);
+        return $dependencies;
     }
 
-    private function autowireDependency(ReflectionParameter $parameter, ContainerLog $log): mixed
+    private function autowireDependency(ReflectionParameter $parameter, mixed $providedValue = null): mixed
     {
         $parameterType = $parameter->getType();
 
         // If the parameter is a built-in type, immediately skip
         // reflection stuff and attempt to give it a default
         // or null value.
         if ($parameterType instanceof ReflectionNamedType && $parameterType->isBuiltin()) {
-            return $this->autowireBuiltinDependency($parameter);
-        }
-
-        // If there are multiple possible types, loop through them
-        // until we hit a match.
-        if ($parameterType instanceof ReflectionUnionType) {
-            foreach ($parameterType->getTypes() as $type) {
-                try {
-                    if ($instance = $this->resolve($type->getName(), $log)) {
-                        return $instance;
-                    }
-                } catch (Throwable) {
-                }
-            }
+            return $this->autowireBuiltinDependency($parameter, $providedValue);
         }
 
-        // If this is a single type, attempt to resolve it.
-        try {
-            return $this->resolve($parameter->getType()->getName(), $log);
-        } catch (Throwable) {
+        // Convert the types to an array regardless, so we can handle
+        // union types and single types the same.
+        $types = ($parameterType instanceof ReflectionUnionType)
+            ? $parameterType->getTypes()
+            : [$parameterType];
+
+        // Loop through each type until we hit a match.
+        foreach ($types as $type) {
+            try {
+                return $this->autowireObjectDependency($type, $providedValue);
+            } catch (Throwable) {
```

Comment:
Add a comment in this empty catch block to clarify that we're just gonna try the next Type in the union.